pkg/apputils/utils.go

```diff
@@ -0,0 +1,191 @@
+)
+
+//go:generate go run github.com/maxbrunsfeld/counterfeiter/v6 -generate
+//counterfeiter:generate . AppFactory
+type AppFactory interface {
+	GetKubeService() (kube.Kube, error)
+	GetAppService(ctx context.Context, name, namespace string) (*app.App, error)
+}
+
+type DefaultAppFactory struct {
+}
+
+func (f *DefaultAppFactory) GetAppService(ctx context.Context, name, namespace string) (*app.App, error) {
+	return GetAppService(ctx, name, namespace)
+}
+
+func (f *DefaultAppFactory) GetKubeService() (kube.Kube, error) {
+	kubeClient, _, kubeErr := kube.NewKubeHTTPClient()
+	if kubeErr != nil {
+		return nil, fmt.Errorf("error creating k8s http client: %w", kubeErr)
+	}
+
+	return kubeClient, nil
+}
+
+func GetLogger() logger.Logger {
+	osysClient := osys.New()
+	return logger.NewCLILogger(osysClient.Stdout())
+}
+
+func getBaseClients() (osys.Osys, flux.Flux, kube.Kube, logger.Logger, error) {
+	osysClient := osys.New()
+	cliRunner := &runner.CLIRunner{}
+	fluxClient := flux.New(osysClient, cliRunner)
+
+	kubeClient, _, kubeErr := kube.NewKubeHTTPClient()
+	if kubeErr != nil {
+		return nil, nil, nil, nil, fmt.Errorf("error creating k8s http client: %w", kubeErr)
+	}
+
+	logger := logger.NewCLILogger(osysClient.Stdout())
+
+	return osysClient, fluxClient, kubeClient, logger, nil
+}
+
+func IsClusterReady() error {
+	logger := GetLogger()
+	kube, _, kubeErr := kube.NewKubeHTTPClient()
+	if kubeErr != nil {
+		return fmt.Errorf("error creating k8s http client: %w", kubeErr)
+	}
+
+	if readyErr := app.IsClusterReady(logger, kube); readyErr != nil {
+		return readyErr
+	}
+
+	return nil
+}
+
+func GetAppService(ctx context.Context, appName, namespace string) (*app.App, error) {
+	osysClient, fluxClient, kubeClient, logger, baseClientErr := getBaseClients()
+	if baseClientErr != nil {
+		return nil, fmt.Errorf("error initializing clients: %w", baseClientErr)
+	}
+
+	appClient, configClient, gitProvider, clientErr := getGitClientsForApp(ctx, appName, namespace)
+	if clientErr != nil {
+		return nil, fmt.Errorf("error getting git clients: %w", clientErr)
+	}
+
+	return app.New(logger, appClient, configClient, gitProvider, fluxClient, kubeClient, osysClient), nil
+}
+
+func GetAppServiceForAdd(ctx context.Context, url, configUrl, namespace string, isHelmRepository bool) (*app.App, error) {
+	osysClient, fluxClient, kubeClient, logger, baseClientErr := getBaseClients()
+	if baseClientErr != nil {
+		return nil, fmt.Errorf("error initializing clients: %w", baseClientErr)
+	}
+
+	appClient, configClient, gitProvider, clientErr := getGitClients(ctx, url, configUrl, namespace, isHelmRepository)
+	if clientErr != nil {
+		return nil, fmt.Errorf("error getting git clients: %w", clientErr)
+	}
+
+	return app.New(logger, appClient, configClient, gitProvider, fluxClient, kubeClient, osysClient), nil
+}
+
+func getGitClientsForApp(ctx context.Context, appName, namespace string) (git.Git, git.Git, gitproviders.GitProvider, error) {
```

Comment:
Idea: the passing around of `git.Git, git.Git, gitproviders.GitProvider` is so common I wonder if a dedicated struct makes sense?

Reply:
Maybe? It's only two functions in one file :man_shrugging:

Reply:
And then they are passed into `app.App`.

Reply:
Chatted f2f, not pull this thread on this PR given the large number of changes.